### DISCUSSION TITLE
Incorporating some of the Functionality of AccuDraw

### DISCUSSION THREAD

**Discussion by Rycha2021** [iTwin/itwinjs-core#3755] [Q&A]
We are interested in understanding whether some of the functionality from AccuDraw can be incorporated into iTwin to draw shapes within a Reality Model? 

We see basic shapes maybe possible from the example library however, can we draw a cylinder, for example, from a piece of equipment to another through a basic UI to click and draw?

Interested to learn if this is available now with V3 or is it on the near term dev roadmap?

Ry

**Comment by bbastings** [comment]
AccuDraw is available and can be used by PrimitiveTools you create to aid in geometry creation. Refer to [AccuDraw](https://www.itwinjs.org/learning/frontend/primitivetools/#accudraw)

**Comment by Rycha2021** [comment]
Perhaps we have misunderstood what AccuDraw does. We understood it can enable the user to draw rudimentary shapes, i.e. cubes / cylinders.
The use case we are wanting to implement is for a user to draw and extrude basic shapes between points the user defines. We have seen the iTwin example of some basic shapes.

The user ease sought is similar to how the 'measure tool' works, i.e. select a point click, select next point click and then the shape is drawn. 

For a little more complex use case, creating an isometric shape with the functionality and ease of AccuDraw's, click on a corner, and extend the shape X/Y and/or Z distance and rotation. I suppose a basic version of Microstation CONNECT 3D. (https://www.youtube.com/watch?v=qO6O5_Pi830)

This video is way too complicated for what we were considering but you get the idea of how we would like to explore if inserting and drawing defined shapes is possible within a reality model for example.

Hope the description above makes sense and interested if this is possible or on the current roadmap.

Ry

**Reply by bbastings** [reply]
What you are looking for is writing a [PrimitiveTool](https://www.itwinjs.org/learning/frontend/primitivetools/). AccuDraw and AccuSnap are drawing aids that a tool can enable so that the user of the tool can more easily specify coordinates/directions/distances, etc.

PrimitiveTools can be created today for constructing and modifying 3d geometry in much that same way as what was shown in the MicroStation video you linked. The [editor](https://github.com/iTwin/itwinjs-core/tree/master/editor) package includes quite a few.

We are currently working on creating a new/better example editing application to demonstrates setting up a briefcase that communicates with a dedicated backend and can push changes back to the hub. Editing an imodel can be demonstrated today using [ui-test-app](https://github.com/iTwin/itwinjs-core/tree/master/test-apps/ui-test-app) however this feature is currently @alpha and it might be better to wait on the new, more focused example.

**Reply by Rycha2021** [reply]
Ok this reads promising and along the track of what we would consider worthwhile.

Alpha implies imminent release i presume? Is there an ETA for this. We would be keen to get this progressing as soon as Alpha is deployed.

Thanks again. 

Ry

**Reply by pmconne** [reply]
"Alpha" is a [release tag](https://www.itwinjs.org/learning/guidelines/release-tags-guidelines/), implying the feature is already released. It describes an API that is experimental, subject to change, and not yet publicly documented. Above, it refers to the APIs in the editor package.
You do not appear to want to edit the contents of iModels, but instead write a tool that creates transient graphics in the context of a reality model. For that, you can use PrimitiveTool, AccuDrawHintBuilder, etc, which are "public", not "alpha", meaning those APIs are documented and guaranteed to remain stable for at least the next major package version.
Did you look at [the article](https://www.itwinjs.org/learning/frontend/primitivetools/) @bbastings linked? At the bottom you will find snippets from a PrimitiveTool that uses AccuDraw to draw line segments. You can adapt it to draw cylinders or whatever else you want.
